docs/auth/oidc.md

```diff
@@ -0,0 +1,234 @@
+
+### The API Factory
+
+A factory is a function that can take some parameters or dependencies and return an
+instance of something, in our case it will be a function that requests some backstage
+APIs and use them to create an instance of an OIDC provider.
+
+Please note that this function only runs (creates the instance) when somewhere else in
+the app you request the DI to give you an instance of the OIDC provider using the API ref
+defined above, and the DI will only run this function the first time, from then on any
+other DI injection will just receive the same instance created the first time, basically
+the instance is cached by the DI library, a singleton.
+
+Let's add our OIDC factory to the APIs array in the `packages/app/src/apis.ts` file:
+
+```diff
++ import { OAuth2 } from '@backstage/core-app-api';
+
+export const apis: AnyApiFactory[] = [
++   createApiFactory({
++    api: oidcAuthApiRef,
++    deps: {
++      discoveryApi: discoveryApiRef,
++      oauthRequestApi: oauthRequestApiRef,
++      configApi: configApiRef,
++    },
++    factory: ({ discoveryApi, oauthRequestApi, configApi }) =>
++      OAuth2.create({
++        discoveryApi,
++        oauthRequestApi,
++        provider: {
++          id: 'oidc',
++          title: 'OIDC provider',
++          icon: () => null,
++        },
++        environment: configApi.getOptionalString('auth.environment'),
++        defaultScopes: [
++          'openid',
++          'profile',
++          'email',
++        ],
++      }),
++  }),
+
+```
+
+Please note we're importing the `OAuth2` class from `@backstage/core-app-api` effectively
+delegating the authentication to it (yes it can handle OIDC as well). Also we're using
+the `oidc` ID to tell `OAuth2` to use the OIDC protocol, and added the default scopes to
+request ID, profile, email and user read permissions.
+
+### The Resolver
+
+Resolvers exist to map user identity from the 3rd party (in this case OIDC provider) to
+the backstage user identity, for a detailed explanation check the [Identity Resolver][1]
+page, it explains how to write a custom resolver as well as linking the built in resolvers
+of backstage.
+
+As an example if you're setting up OIDC provider with Microsoft, you could use the built
+in Microsoft resolvers, or create one yourself in `packages/backend/src/plugins/auth.ts`:
+
+```diff
+import {
+  DEFAULT_NAMESPACE,
++ stringifyEntityRef,
+} from '@backstage/catalog-model';
+
+export default async function createPlugin(
+  env: PluginEnvironment,
+): Promise<Router> {
+  return await createRouter({
+    logger: env.logger,
+    config: env.config,
+    database: env.database,
+    discovery: env.discovery,
+    tokenManager: env.tokenManager,
+    providerFactories: {
+      ...defaultAuthProviderFactories,
++     // oidc: providers.oidc.create({
++     //   signIn: {
++     //     resolver:
++     //       providers.microsoft.resolvers.emailMatchingUserEntityAnnotation() as any,
++     //   },
++     // }),
```

Comment:
I think we could drop this commented out code?

Reply:
Tought the example of using another provider's resolver could be useful.

Reply:
I'll remove it, it might be a bad practice to use the resolvers of other providers given the typings will not match, even if the result is the same for most of them.